src/libostree/ostree-repo-checkout.c

```diff
@@ -329,25 +358,54 @@ checkout_file_hardlink (OstreeRepo                          *self,
   else if (allow_noent && errno == ENOENT)
     {
     }
-  else if (errno == EEXIST && options->overwrite_mode == OSTREE_REPO_CHECKOUT_OVERWRITE_ADD_FILES)
+  else if (errno == EEXIST)
     {
-      /* In this mode, we keep existing content.  Distinguish this case though to
-       * avoid inserting into the devino cache.
-       */
-      ret_result = HARDLINK_RESULT_SKIP_EXISTED;
-    }
-  else if (errno == EEXIST && options->overwrite_mode == OSTREE_REPO_CHECKOUT_OVERWRITE_UNION_FILES)
-    {
-      /* Idiocy, from man rename(2)
-       *
-       * "If oldpath and newpath are existing hard links referring to
-       * the same file, then rename() does nothing, and returns a
-       * success status."
-       *
-       * So we can't make this atomic.
-       */
-      (void) unlinkat (destination_dfd, destination_name, 0);
-      goto again;
+      if (options->overwrite_mode == OSTREE_REPO_CHECKOUT_OVERWRITE_ADD_FILES)
+        {
+          /* In this mode, we keep existing content.  Distinguish this case though to
+           * avoid inserting into the devino cache.
+           */
+          ret_result = HARDLINK_RESULT_SKIP_EXISTED;
+        }
+      else if (options->overwrite_mode == OSTREE_REPO_CHECKOUT_OVERWRITE_UNION_FILES)
+        {
+          /* Idiocy, from man rename(2)
+           *
+           * "If oldpath and newpath are existing hard links referring to
+           * the same file, then rename() does nothing, and returns a
+           * success status."
+           *
+           * So we can't make this atomic.
+           */
+          (void) unlinkat (destination_dfd, destination_name, 0);
+          goto again;
+        }
+      else if (options->overwrite_mode == OSTREE_REPO_CHECKOUT_OVERWRITE_DISJOINT_UNION_FILES)
+        {
+          /* In this mode, we error out on EEXIST *unless* the files are already
+           * hardlinked, which is what rpm-ostree wants for package layering.
+           * https://github.com/projectatomic/rpm-ostree/issues/982
+           */
+          struct stat src_stbuf;
+          if (!glnx_fstatat (srcfd, loose_path, &src_stbuf,
+                             AT_SYMLINK_NOFOLLOW, error))
+            return FALSE;
+          struct stat dest_stbuf;
+          if (!glnx_fstatat (destination_dfd, destination_name, &dest_stbuf,
+                             AT_SYMLINK_NOFOLLOW, error))
+            return FALSE;
+          const gboolean is_identical =
+            (src_stbuf.st_dev == dest_stbuf.st_dev &&
```

Comment:
Hey, this line is checking for if file from the old src and the new destination is on the same device and having same file serial number right? Then if both are same, we can conclude two files are identical? But, it also seems a bit weird to me because to me identical also kinda applies when two files share the same name and exact same content. 

I am not sure if this case will happen in ostree checkout case, but let's say if we were checking out `/usr/share/xz/file` to the ostree and there is a file already existing at the same location (`/usr/share/xz/file`) , with the same content, will this implementation still catches that? 

Again, I am not very familiar with the codebase, so. this might be a misinterpretation of the code, sorry if that happens, and thanks in advance!  :)

Reply:
Hmm, I guess we could also just checksum the file to determine if it's identical. Though I'm not sure what the chances of them having the same content + file info but somehow *not* being from a checkout. (E.g. ostree canonicalizes timestamps to epoch, which would be odd in any other situations).

Reply:
ah it seems my question was poorly worded :(, apologies. 

So, what I actually meant was that do we also account for the case that two files have same file content,path , but different st_ino. (Not sure if that is possible to happen for checking out files).

But if that situation exists, should we also consider those two files to be identical?

Reply:
If checkouts aren't hardlinked, a lot (not all, but most) of the value of ostree drops away.  So I'm not sure it's worth thinking too much about the case where we're not hardlinking.  Right?

Reply:
Latest fixup ⬇️  makes the non-hardlinking disjoint union path an explicit error.

Reply:
Ah I believe I thought it wrong before. It makes more sense now. Thanks for the clarification :)